[SecurityBundle] Empty line starting with dash under "access_control" causes all rules to be skipped

monteiro opened the pull request:
| Q             | A
| ------------- | ---
| Branch?       | 4.4
| Bug fix?      | yes
| New feature?  | no <!-- please update src/**/CHANGELOG.md files -->
| Deprecations? | no <!-- please update UPGRADE-*.md and src/**/CHANGELOG.md files -->
| Tickets       | Fix #40235 ... <!-- prefix each issue number with "Fix #", no need to create an issue if none exist, explain below instead -->
| License       | MIT

When the IDE by mistake puts an empty line in `access_control` in security.yaml there is no warning that we have an empty row, making the rest of routes defined, to be ignored and possible to be accessed by anyone that can authenticate no matter the role.

# How to reproduce the issue

- git clone git@github.com:monteiro/symfony-issue-40235.git
- composer install
- symfony server:start
- open 127.0.0.1:8000/admin with username: "john_user" and password "123456"
- Since that user has only ROLE_USER should not be able to access the route... but because there is an empty line in "access_control" in `security.yaml`, "by mistake" it is possible to access the protected `ROLE_ADMIN` route.

monteiro committed:
[SecurityBundle] Empty line starting with dash under "access_control" causes all rules to be skipped

derrabus commented on src/Symfony/Bundle/SecurityBundle/Tests/DependencyInjection/SecurityExtensionTest.php:
```diff
@@ -415,6 +415,31 @@ public function testSwitchUserWithSeveralDefinedProvidersButNoFirewallRootProvid
 
+    public function testInvalidAccessControlWithEmptyRow()
+    {
+        $this->expectException(InvalidConfigurationException::class);
+        $this->expectExceptionMessage('There is an empty access control attributes set.');
```
Those to lines should be moved downwards, right above the line that we expect to trigger the exception.

monteiro replied:
Done!

derrabus reviewed (approved):
Looks good, I'm merely nit-picking here. 🙂

wouterj commented:
Thank you for creating this PR including tests and a reproducer, @monteiro! That's perfect!

After looking at the changes, I'm wondering whether we qualify this as a BC break. More precisely, people may have been using the "empty attributes" feature to exclude some routes from the access control (e.g. a login form). However, this was never documented behavior.

Let's wait a bit for more comments from the core team whether or not this is a BC break. If we qualify it as such, I think it makes all sense to deprecate this in 5.x instead (and introducing this exception in 6.0). `IS_AUTHENTICATED_ANONYMOUSLY` or `PUBLIC_ACCESS` should be used instead of an empty array of attributes imho.

derrabus commented:
> After looking at the changes, I'm wondering whether we qualify this as a BC break.

I thought about that as well and could live with going through the deprecation process for that.

> More precisely, people may have been using the "empty attributes" feature to exclude some routes from the access control (e.g. a login form). However, this was never documented behavior.

My current understanding of the behavior (please correct me if I'm wrong) is that the empty line simply matches every request without applying any access restrictions. The result is effectively that any subsequent line would become unreachable. That behavior is dangerously surprising imho.

Can we come up with a scenario where this change breaks an application that is not dangerously misconfigured?

wouterj commented:
> Can we come up with a scenario where this change breaks an application that is not dangerously misconfigured?

```yaml
access_control:
  - { path: ^/login }
  - { path: ^/, roles: ROLE_USER }
```

I haven't verified it, but as far as I understand the code, this yaml file is equal to doing `- { path: ^/login, roles: IS_AUTHENTICATED_ANONYMOUSLY }`.

The diff is checking `$attributes`, which are the `roles` + `allow_if` statements from access control.

derrabus commented:
@wouterj With your example, the exception in this PR should not be triggered. The case that is caught here is this one:

```YAML
access_control:
  -
  - { path: ^/, roles: ROLE_USER }
```

wouterj commented:
@derrabus as indicated with my last sentence, this case *is* caught by this PR, as the check is done over `$attributes`, which only contains the values of `roles` and `allow_if`. So maybe we should do this check over `$access`? (not sure how that works, as most options have a default value of `null`.

derrabus commented:
> @derrabus as indicated with my last sentence, this case _is_ caught by this PR

Okay, that's indeed an issue then.

derrabus reviewed (changes requested):
We need to make sure that @wouterj's example still works.

xabbuh commented on src/Symfony/Bundle/SecurityBundle/DependencyInjection/SecurityExtension.php:
```diff
@@ -191,6 +191,10 @@ private function createAuthorization(array $config, ContainerBuilder $container)
                 $attributes[] = $this->createExpression($container, $access['allow_if']);
             }
 
+            if (!$attributes) {
```
What about instead checking that at least one of the keys in `$access` is not `null`?

xabbuh replied:
or the empty array in case of `method`, `ips` and `roles`

monteiro commented:
Following @xabbuh advice I fixed this specific edge case and added @wouterj example to the test, so we are sure it is not breaking old behaviour. Let me know what you think.

Thanks for the help!

wouterj commented on src/Symfony/Bundle/SecurityBundle/DependencyInjection/SecurityExtension.php:
```diff
@@ -191,6 +191,14 @@ private function createAuthorization(array $config, ContainerBuilder $container)
+            }));
+
+            if ($isEmptyAccess) {
+                throw new InvalidConfigurationException('There is an empty access control attributes set.');
```
```suggestion
                throw new InvalidConfigurationException('One or more access control items are empty. Did you accidentally add lines only containing a "-" under "security.access_control"?');
```

wouterj commented on src/Symfony/Bundle/SecurityBundle/DependencyInjection/SecurityExtension.php:
```diff
@@ -191,6 +191,14 @@ private function createAuthorization(array $config, ContainerBuilder $container)
                 $attributes[] = $this->createExpression($container, $access['allow_if']);
             }
 
+            $isEmptyAccess = 0 === \count(array_filter($access, function ($v) {
+                return !empty($v);
+            }));
+
+            if ($isEmptyAccess) {
```
```suggestion
            $emptyAccess = array_filter($access, function ($v) {
                return array_filter($v);
            });

            if ($emptyAccess) {
```

I'm not sure, but I think this suggestion works. I believe we try to avoid `empty()` where possible in Symfony.

wouterj reviewed (commented):
Maybe I was a bit too fast with my approval. I'm not 100% sure if my suggestions make sense.

monteiro commented:
Let me know if there is anything I can do to help here.

derrabus commented:
Thank you @monteiro.

monteiro commented:
Thanks to you @derrabus
